File: common/python/primality.py
import math
from functools import lru_cache
from collections import defaultdict
from itertools import combinations


@lru_cache(maxsize=None)
def is_prime(num):
    if num == 2:
        return True
    if num % 2 == 0:
        return False
    top = int(math.floor(math.sqrt(num)))
    for i in range(3, top + 1, 2):
        if num % i == 0:
            return False
    return True


def prime_factors(num):
    if num <= 1:
        return defaultdict(int)

    prime_divs = defaultdict(int)

    trial = 2
    num_ = num
    top = int(math.sqrt(num_))

    while True:
        if trial > top:
            d = defaultdict(int)
            d[num_] = 1
            return d

        if num_ % trial == 0:
            parent = prime_factors(num_ // trial)
            parent[trial] += 1
            return parent
        else:
            if trial == 2:
                trial += 1
            else:
                trial += 2

@lru_cache(maxsize=None)
def factors(num):
    if num <= 1:
        return defaultdict(int)

    divs = set()

    trial = 2
    num_ = num
    top = num // 2

    while True:
        if trial > top:
            return set([1])

        if num_ % trial == 0:
            parent = factors(num_ // trial)
            all_ = set()
            for p in parent:
                all_.add(p)
                all_.add(p * trial)
            return all_
        else:
            if trial == 2:
                trial += 1
            else:
                trial += 2


def prime_sieve(upto):
    nums = set(range(2, upto+1))
    num = 2
    while True:
        if num in nums:
            mult = 2
            while True:
                try:
                    nums.remove(mult * num)
                except KeyError:
                    pass
                if mult * num > upto:
                    break
                mult += 1

        elif num > upto:
            return nums

        num += 1


def phi(n):
    """
    phi(n) = n*sum_{p|n} (1 - 1/p)
    phi(n) = n*sum_{p|n} (p-1)/p
    """
    top = n
    bot = 1

    pfactors = prime_factors(n)

    for p, count in pfactors.items():
        top *= (p - 1)
        bot *= p
    return top // bot


if __name__ == '__main__':
    print(prime_sieve(1000000))

File: common/java/primality/Primality.java
package primality;

import java.util.HashMap;
import java.util.List;
import java.util.HashSet;
import java.util.Arrays;

public class Primality {

	public static boolean isPrime(int val) {
		if(val == 2) {
			return true;
		}

		if((val < 2) || (val % 2 == 0)) {
			return false;
		}

		int upto = (int) Math.round(Math.sqrt(val));

		for(int i = 3; i <= upto; i+=2) {
			if(val % i == 0) {
				return false;
			}
		}

		return true;
	}

	public static HashMap<Integer, Integer> primeFactors(int num) {
		HashMap<Integer, Integer> prime_divs = new HashMap<Integer, Integer>();

		int trial = 2;
		int num_ = num;
		int top = (int) Math.round(Math.sqrt(num_));

		while(true) {
			if(trial > top) {
				HashMap<Integer, Integer> d = new HashMap<Integer, Integer>();
				d.put(num_, 1);
				return d;
			} else if(num_ % trial == 0) {
				HashMap<Integer, Integer> parent = primeFactors(num_ / trial);
				Integer current = parent.get(trial);
				if(current == null) {
					current = 0;
				}
				parent.put(trial, current + 1);
				return parent;
			} else {
				if(trial == 2) {
					++trial;
				} else {
					trial += 2;
				}
			}
		}
	}

	/*
	public static ArrayList<ArrayList<Integer>> factorisations(int num) {
		//ArrayList<Integer> skip = new ArrayList<Integer>();
		return factorisations(num, ski);
	}
	*/

	public static HashSet<HashMap<Integer, Integer>> factorisations(int num) {
		if(num == 1) {
			HashSet<HashMap<Integer, Integer>> l = new HashSet<HashMap<Integer, Integer>>();
			l.add(new HashMap<Integer, Integer>());
			return l;
		}

		HashSet<HashMap<Integer, Integer>> fs = new HashSet<HashMap<Integer, Integer>>();
		for(int i = 2; i <= num / 2; ++i) {
			if(num % i != 0) {
				continue;
			}

			for(HashMap<Integer, Integer> l : factorisations(num / i)) {
				if(l.get(i) == null) {
					l.put(i, 1);
				} else {
					l.put(i, l.get(i) + 1);
				}
				fs.add(l);
			}
		}

		for(HashMap<Integer, Integer> l : factorisations(1)) {
			if(l.get(num) == null) {
				l.put(num, 1);
			} else {
				l.put(num, l.get(num) + 1);
			}
			fs.add(l);
		}

		return fs;
	}
}
